### DISCUSSION TITLE
How to show a List of objects

### DISCUSSION THREAD

**Discussion by mkj-stonehill** [AvaloniaUI/Avalonia#9927] [Q&A] [answered]
I'm new to Avalonia and MVVM (and it's been about a decade since I did any WPF).  My application needs to show a list of objects (in a horizontal StackPanel), where the List is created when the application starts up (each item represents an instance of a hardware module which may or may not be present; there can be from 1 to 6 of them).

I've defined a class to represent the module, and a view which is a UserControl to present to relevant data from the module class.  The MainWindowViewModel constructor detects which module(s) are present, creates a module object for each one, and adds it to a list.  What I can't figure out is the magic XAML to display the elements of that list as items in the StackPanel on the main window.

If someone can point me in the right direction, I would be forever grateful.

**Comment by lukaszossowski** [comment, marked as answer]
If you are doing it with MVVM you should create a HardwareModuleViewModel which exposes your model's (the hardware module) properties, implementing INotifyPropertyChanged. Then you should create an ObservableCollection of type HardwareModuleViewModel in your MainWindowViewModel and use it to display data in the UserControl. You can display them using different controls - a ListBox, a ContentControl, a DataGrid etc. 

Using a ListBox, for example, simply bind the observable collection to lists's Items (In XAML it will be something like:
```

		<ListBox Items="{Binding HardwareModuleVMs}">
			(...)
		</ListBox>
```
Inside the listbox you will have to define a data template to display your data properly. If you leave it blank it will just show a list of "HardwareModuleViewModel" because it will just display the type name.

I'm writing this from memory, so this might not be perfect but I think it should be enough to point you to the right direction. Another good direction would be to google how to do that in WPF, you will get many results and they will also aply to Avalonia, sometimes with minor changes required.

**Comment by mkj-stonehill** [comment]
The answer from @lukaszossowski did indeed point me in the right direction (especially the advice to google how to do it in WPF; I didn't realize how closely related Avalonia and WPF were).  For completeness, here's what I did:

1. The class implementing the model (in my case, the `HardwareModule` class) needs to inherit from `INotifyPropertyChanged`, which means you need to add a member `public event PropertyChangedEventHandler? PropertyChanged;`.  Then, when any property changes, you need to call `PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(X)));`, where X is the property being changed.

2. Add an observable collection of the model to the ViewModel: `public ObservableCollection<Models.HardwareModule> Modules { get; }` (I create and add the HardwareModule items in the constructor of the ViewModel, that's why there's no setter).

3. In the View, add the following (this is for adding the items in the collection to a StackPanel):
```
    <!--This binds to the observable collection in the ViewModel-->
    <ItemsControl Items="{Binding Modules}">

      <!--Use a StackPanel to display all the modules-->
      <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
          <StackPanel />
        </ItemsPanelTemplate>
      </ItemsControl.ItemsPanel>

      <!--This defines the layout for each item (i.e. each module)-->
      <ItemsControl.ItemTemplate>
        <DataTemplate>
           <!--The controls to display each item; bind them to properties on the HardwareModule class-->
        </DataTemplate>
      </ItemsControl.ItemTemplate>
    </ItemsControl>
```
Hope that helps anyone who comes across this question in the future.

**Reply by timunie** [reply]
marked it as answer for others to find it faster.

**Reply by Damglador** [reply]
For whoever stumbles upon this in the future, in Avalonia 11+ replace `Items={Binding ...}` with `ItemsSource={Binding ..}`

**Comment by mysteryname** [comment]
Hello, 

I had some trouble getting the above to work in AvaloniaUI 11 Compiling with a `ObservableCollection<T>` where `<T>` is a custom class. Try changing the above examples to the following:

```
    <!--This binds to the observable collection in the ViewModel-->
    <ItemsControl ItemsSource="{Binding Modules}">
    <!--The above line has changed to "ItemsSource" rather than "Items"-->

      <!--Use a StackPanel to display all the modules-->
      <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
          <StackPanel />
        </ItemsPanelTemplate>
      </ItemsControl.ItemsPanel>

      <!--This defines the layout for each item (i.e. each module)-->
      <ItemsControl.ItemTemplate>
        <DataTemplate>
           <!--The controls to display each item; bind them to properties on the HardwareModule class-->
        </DataTemplate>
      </ItemsControl.ItemTemplate>
    </ItemsControl>
```

This assumes that the `Modules` is a class similar to the following: 

```csharp
public class Modules
{
  public string Message { get; set; }
  public int MessageId { get; set; }
}
```
